### ISSUE TITLE
on_validation_batch_end is not called when Loss is NaN

### ISSUE THREAD

**Issue by PeiyuanQi** [Lightning-AI/pytorch-lightning#20999]
### Bug description

Problem: on_validation_batch_end is not called when Loss is NaN.

How to reproduce: create a model with loss as NaN invalidation, and then on_validation_batch_end won't be called.

### What version are you seeing the problem on?

v2.5

### How to reproduce the bug

```python

```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.5.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

**Comment by deependujha** [comment]
Hi @PeiyuanQi ,
Thank you for opening the issue, I'm currently looking into it.

As a suggestion, it would be very helpful if you could share a [lightning studio](https://lightning.ai/) that reproduces the issue. This would greatly assist us in triaging and resolving it more efficiently.

Thanks again!
Deependu

**Comment by deependujha** [comment]
Hi @PeiyuanQi,

I ran a test with a model that produces NaN during validation, and I’m seeing that `on_validation_batch_end` **is** being called as expected (screenshot attached).

<details>
  <summary>Here's my code</summary>

```python
import math
import argparse

import lightning as L
import torch
import torch.nn as nn
from lightning.pytorch.callbacks import ModelCheckpoint, RichProgressBar
from torch.utils.data import DataLoader, TensorDataset


callbacks = [
    RichProgressBar(),
]


class SimpleModel(L.LightningModule):
    def __init__(self):
        super().__init__()
        self.linear = nn.Linear(10, 1)
        self.loss_fn = nn.MSELoss()

    def forward(self, x):
        return self.linear(x)

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = self.loss_fn(y_hat, y)
        self.log("train_loss", loss)
        return loss

    def validation_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = self.loss_fn(y_hat, y)
        self.log("val_loss", loss)
        return torch.tensor(math.nan)

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.01)


def create_random_data(num_samples=1000):
    x = torch.randn(num_samples, 10)
    y = torch.randn(num_samples, 1)
    return x, y


if __name__ == "__main__":
    x_train, y_train = create_random_data(1000)
    x_val, y_val = create_random_data(200)

    train_dataset = TensorDataset(x_train, y_train)
    val_dataset = TensorDataset(x_val, y_val)

    train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True, num_workers=1, persistent_workers=True)
    val_loader = DataLoader(val_dataset, batch_size=32, num_workers=1, persistent_workers=True)

    model = SimpleModel()

    trainer = L.Trainer(
        max_epochs=3,
        enable_progress_bar=True,
        log_every_n_steps=1,
        val_check_interval=0.5,
        callbacks=callbacks,
        devices=[0],
    )

    trainer.fit(model, train_loader, val_loader)
```
</details>

<img width="1440" height="900" alt="Image" src="https://github.com/user-attachments/assets/4fcada3b-46ac-4a95-8f7f-36f0df47df8b" />

It’s possible there might be a difference in how the NaN is being introduced or handled in your setup. If you could share a minimal [Lightning Studio](https://lightning.ai/) that reproduces the issue, I’d be happy to investigate further.

Best,
Deependu

**Comment by deependujha** [comment]
Hi @PeiyuanQi , just checking in to see if you had a chance to create a minimal reproducible example for this issue. That would help us investigate further.

**Comment by PeiyuanQi** [comment]
ah sry, just realize we didn't reply to this thread, based on further investigation we think the problem is on the app implication side, we could close this thread.
